Make SER_GETHASH implicit for CHashWriter and SerializeHash

Empact opened the pull request:
Most invocations of `CHashWriter` use `SER_GETHASH` and version 0, this allows for eliding those values, ~and removes `SER_GETHASH` as a type, because it functions simply as "has no external destination" in practice.~

~`SerializeHash` basically existed due to the overhead of CHashWriter construction, now that that is minimized, it's unnecessary.~

Empact committed:
Drop nType argument from SerializeHash

All callers either relied on the default value of SER_GETHASH or provided
SER_GETHASH.

Empact committed:
Add convenience constructors for CHashWriter

Empact commented on src/primitives/block.cpp:
```diff
@@ -12,7 +12,7 @@
 uint256 CBlockHeader::GetHash() const
 {
-    return SerializeHash(*this);
+    return (CHashWriter(PROTOCOL_VERSION) << *this).GetHash();
```
Note this was the only call that relied on the default `SerializeHash` `nVersion` value.

Empact commented:
Withdrawing this pending #10785. Will re-open after.

Empact commented:
Good to know, thanks!

sipa commented:
I don't understand "scripted-diff: Drop SER_GETHASH"; it changes `!(s.GetType() & SER_GETHASH)` to `s.GetType()`. That means that what used to be the branch for DISK/NETWORK will end up being run just for DISK, and what used to be the branch for GETHASH will end up being run for NETWORK and GETHASH.

sipa commented:
Ah of course; I missed the `(1 << ...)` around it.

This isn't very readable code though. Would you mind keeping `SER_GETHASH` as a constant (perhaps just defined as 0), and explicit comparisons with it (rather than implicit conversion of enum->int->bool to mean nonzero)?

Empact commented:
@sipa Fair enough, I dropped that commit. Def more straight-forward now.

sipa commented on src/hash.h:
```diff
@@ -122,8 +122,8 @@ class CHashWriter
 
-    CHashWriter() : CHashWriter(SER_GETHASH, 0) {}
-    CHashWriter(int nVersionIn) : CHashWriter(SER_GETHASH, nVersionIn) {}
+    CHashWriter() : CHashWriter(0, 0) {}
```
Can you use SER_NETWORK instead here? 0 is not a valid serialization type, even if it's not actually used anywhere.

Empact replied:
Reviewing calls to `GetType` after these changes, it seems like there are only 2 calls that affect the output, both testing `GetType` for `SER_DISK` in `CAddress::SerializationOp`. Is it possible to remove those? If so we could get right of `GetType` entirely, which seems like a big win.

If not, it seems like SER_NETWORK would not cause negative effects, but it would be a bit misleading for those cases where the serialization is not propagated to the network.

Empact replied:
Relatedly: #13558

Empact replied:
As of #13560, I'm pretty confident that `GetType` can be removed across the board.

Empact replied:
Switched to `SER_NETWORK` - it's equivalent to `0` now as only `SER_DISK` has conditional behavior associated with it.

practicalswift commented on src/hash.h:
```diff
@@ -122,6 +122,8 @@ class CHashWriter
 public:
 
+    CHashWriter() : CHashWriter(0, 0) {}
+    CHashWriter(int nVersionIn) : CHashWriter(0, nVersionIn) {}
```
Should be `explicit`? :-)

practicalswift replied:
```
2018-09-22 22:13:42 cpplint(pr=13462): src/hash.h:126:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
```

practicalswift commented on src/hash.h:
```diff
@@ -122,6 +122,8 @@ class CHashWriter
 public:
 
+    CHashWriter() : CHashWriter(0, 0) {}
+    CHashWriter(int nVersionIn) : CHashWriter(0, nVersionIn) {}
```
This should be `explicit`? :-)

Empact commented:
In case the SER_GETHASH removal was complicating things, I reoriented this around simply making SER_GETHASH implicit in the obviously hash-specific contexts.

practicalswift commented on src/hash.h:
```diff
@@ -123,6 +123,8 @@ class CHashWriter
 public:
 
+    CHashWriter() : CHashWriter(SER_GETHASH, 0) {}
+    CHashWriter(int nVersionIn) : CHashWriter(SER_GETHASH, nVersionIn) {}
```
Should be `explicit`?

Empact replied:
Indeed, fixed.

Empact committed:
scripted-diff: Use new CHashWriter constructors

-BEGIN VERIFY SCRIPT-
sed -i -E "s/CHashWriter\(SER_GETHASH, 0\)(.[^{])/CHashWriter()\1/" $(git grep -l 'CHashWriter')
sed -i -E "s/CHashWriter (.+)\(SER_GETHASH, 0\)/CHashWriter \1/" $(git grep -l 'CHashWriter')
sed -i -E "s/CHashWriter (.+)\(SER_GETHASH, /CHashWriter \1(/" $(git grep -l 'CHashWriter')
-END VERIFY SCRIPT-

Empact commented:
Rebased, refined, how about another look?

Empact commented:
Rebased

laanwj commented:
This has been open for two years ! we probably should get to either merging it if it is worth doing, or decide not to, but not keep @Empact rebasing it forever.

Empact committed:
Include serialize.h wherever SER_GETHASH is referenced

Empact commented:
Hi @laanwj - I took 2020 (and then some) off, but it's good to be back at it. As you note, this is old work and may not be worthwhile but here are my thoughts:

In these contexts, `SER_GETHASH` is redundant/noise. It's a matter of interpretation whether `CHashWriter(0)` is more informative than `CHashWriter` - I tend to think it does not convey information so should be elided for readability.

It's a bit striking when you consider that of 39 references to `SER_GETHASH` in `master`, 30 of them are content-free/noise. This removes those. The original PR was to remove `SER_GETHASH` entirely, but I walked that back to this more modest change.

As the bulk of the changes are in the form of a scripted-diff covering the specific cases, I look at this as a low-risk improvement to readability - refinement/readability being important to the long-term health of any software project.

If the answer to this is "meh," though, fair enough.

practicalswift commented:
Personally I don't have any strong opinions about this PR but I do have strong opinions about another thing: very glad to have you back @Empact! Warm welcome back as a contributor! :)

>  I took 2020 (and then some) off, but it's good to be back at it.

laanwj commented:
Yes, welcome back!

> If the answer to this is "meh," though, fair enough.

I hope we can get some people who are more familiar with this specific code to review it.
I mean if @sipa is okay with it we should go ahead with it.

jonatack commented:
Approach ACK on this simplification, code looks reasonable on first read.

jonatack reviewed (approved):
At nearly four years old, this may be the pull that has been open the longest so far.

ACK c05246bf00bf8ffa5e73ed739a5001dc3639830c code review, clean rebase to master, clean debug build, ran unit tests

Empact commented:
Incidentally, @jonatack do you happen to know the situation with Cirrus-CI? I'd like to retry [this build](https://cirrus-ci.com/build/6283528743682048), but I don't have the option to do so on the site.

hebasto commented:
> Incidentally, @jonatack do you happen to know the situation with Cirrus-CI? I'd like to retry [this build](https://cirrus-ci.com/build/6283528743682048), but I don't have the option to do so on the site.

Restarted.

MarcoFalke commented:
Not sure on this. I think it might be cleaner to remove ser-type and ser-version completely. See https://github.com/bitcoin/bitcoin/issues/19477#issuecomment-1147421608

MarcoFalke commented:
See https://github.com/bitcoin/bitcoin/pull/25331

MarcoFalke commented:
Can this be closed now?
